python/tests/ert/analysis/test_analysis_module.py

```diff
@@ -73,3 +74,87 @@ def test_create_internal(self):
         mod = AnalysisModule( self.rng , name = "STD_ENKF" )
 
 
+    def test_initX_enkf_linalg_lowrankCinv(self):
+        """Test AnalysisModule.initX with EE=False and GE=False"""
+        mod = AnalysisModule( self.rng , name = "STD_ENKF" )
+        A, S, R, dObs, E, D = self._n_identity_mcs()
+        self.assertFalse(mod.getBool('USE_EE'))
+        self.assertFalse(mod.getBool('USE_GE'))
+
+        elt_a, elt_b = 1.222, -0.111
+        vals = (elt_a, elt_b, elt_b,
+                elt_b, elt_a, elt_b,
+                elt_b, elt_b, elt_a)
+        expected = self.construct_matrix(3, vals)
+
+        X = mod.initX(A, S, R, dObs, E, D)
+        self._matrix_close(X, expected)
+
+    def test_initX_enkf_linalg_lowrank_EE(self):
+        """Test AnalysisModule.initX with EE=True and GE=False"""
+        mod = AnalysisModule( self.rng , name = "STD_ENKF" )
+        A, S, R, dObs, E, D = self._n_identity_mcs()
+        mod.setVar('USE_EE', True)
+        self.assertTrue(mod.getBool('USE_EE'))
+        self.assertFalse(mod.getBool('USE_GE'))
+
+        elt_a, elt_b = 1.33, -0.167
+        vals = (elt_a, elt_b, elt_b,
+                elt_b, elt_a, elt_b,
+                elt_b, elt_b, elt_a)
+        expected = self.construct_matrix(3,  vals)
+        X = mod.initX(A, S, R, dObs, E, D)
+        self._matrix_close(X, expected)
+
+    def test_initX_subspace_inversion_algorithm(self):
+        """Test AnalysisModule.initX with EE=True and GE=True, the subspace inversion algorithm"""
+        mod = AnalysisModule( self.rng , name = "STD_ENKF" )
+        A, S, R, dObs, E, D = self._n_identity_mcs()
+
+        mod.setVar('USE_EE', True)
+        mod.setVar('USE_GE', True)
+        self.assertTrue(mod.getBool('USE_EE'))
+        self.assertTrue(mod.getBool('USE_GE'))
+
+        vals = ( 1.39,  -0.111, -0.278,
+                -0.111,  1.39,  -0.278,
+                 -0.278, -0.278,  1.56 )
+        expected = self.construct_matrix(3, vals)
+        X = mod.initX(A, S, R, dObs, E, D)
+        self._matrix_close(X, expected)
+
+    def construct_matrix(self, n, vals):
+        """Constructs n*n matrix with vals as entries"""
+        self.assertEqual(n*n, len(vals))
+        m = Matrix(n,n)
+        idx = 0
+        for i in range(n):
+            for j in range(n):
+                m[(i,j)] = vals[idx]
+                idx += 1
+        return m
+
+    def _n_identity_mcs(self, n=6,s=3):
+        """return n copies of the identity matrix on s*s elts"""
+        return tuple([self._identity_mx(s) for i in range(n)])
+
+    def _identity_mx(self,s=3):
```

Comment:
Instead of this test special casing I would like to get a first class method to create a diagonal matrix; i.e. `matrix_alloc_identity( int size )` in C and a corresponding classmethod in Python.